File: algorithms/0424-longest-repeating-character-replacement/solution.py
# Longest Repeating Character Replacement - sliding window + counts.
# Time O(n), Space O(1) for A-Z (or alphabet size).

class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        count = [0] * 26
        left = 0
        best = 0
        max_freq = 0
        for right, ch in enumerate(s):
            idx = ord(ch) - ord("A")
            count[idx] += 1
            max_freq = max(max_freq, count[idx])
            while right - left + 1 - max_freq > k:
                count[ord(s[left]) - ord("A")] -= 1
                left += 1
            best = max(best, right - left + 1)
        return best

File: algorithms/0424-longest-repeating-character-replacement/Solution.java
// Longest Repeating Character Replacement - sliding window + counts.
// Time O(n), Space O(1) for A-Z (or alphabet size).

class Solution {
    public int characterReplacement(String s, int k) {
        int[] count = new int[26];
        int left = 0;
        int best = 0;
        int maxFreq = 0;
        for (int right = 0; right < s.length(); right++) {
            char ch = s.charAt(right);
            count[ch - 'A']++;
            maxFreq = Math.max(maxFreq, count[ch - 'A']);
            while (right - left + 1 - maxFreq > k) {
                count[s.charAt(left) - 'A']--;
                left++;
            }
            best = Math.max(best, right - left + 1);
        }
        return best;
    }
}
